src/components/repeat/repeat.component.tsx

```diff
@@ -1,54 +1,52 @@
-import React, { useCallback, useEffect, useId, useState } from 'react';
-import { FormGroup, Button } from '@carbon/react';
-import { Add, TrashCan } from '@carbon/react/icons';
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import { FormGroup } from '@carbon/react';
 import { useFormikContext } from 'formik';
 import { useTranslation } from 'react-i18next';
-import { type FormField, type FormFieldProps } from '../../types';
+import { type FormField, type FormFieldProps, type RenderType } from '../../types';
 import { evaluateAsyncExpression, evaluateExpression } from '../../utils/expression-runner';
-import ObsGroup from '../group/obs-group.component';
 import { isEmpty } from '../../validators/form-validator';
 import styles from './repeat.scss';
-import { cloneObsGroup } from './helpers';
+import { cloneRepeatField } from './helpers';
 import { FormContext } from '../../form-context';
+import { getFieldControlWithFallback } from '../section/helpers';
+import { clearSubmission } from '../../utils/common-utils';
+import RepeatControls from './repeat-controls.component';
 
-export const showAddButton = (limit: string | number, counter: number) => {
-  const repeatLimit = Number(limit);
-  return !Number.isNaN(repeatLimit) && repeatLimit > 0 ? counter < repeatLimit : true;
+const renderingByTypeMap: Record<string, RenderType> = {
+  obsGroup: 'group',
+  testOrder: 'select',
 };
 
-const Repeat: React.FC<FormFieldProps> = ({ question, onChange }) => {
+const Repeat: React.FC<FormFieldProps> = ({ question, onChange, handler }) => {
   const { t } = useTranslation();
-  const id = useId();
-
-  const { fields: allFormFields, encounterContext, obsGroupsToVoid, formFieldHandlers } = React.useContext(FormContext);
-  const { values, setValues } = useFormikContext();
-  const [counter, setCounter] = useState(1);
-  const [obsGroups, setObsGroups] = useState([]);
+  const isGrouped = useMemo(() => question.questions?.length > 1, [question]);
+  const { fields: allFormFields, encounterContext } = React.useContext(FormContext);
+  const { values, setFieldValue } = useFormikContext();
+  const [counter, setCounter] = useState(0);
+  const [rows, setRows] = useState([]);
+  const [fieldComponent, setFieldComponent] = useState(null);
 
   useEffect(() => {
-    const groups = allFormFields.filter(
-      (field) => field.questionOptions.concept === question.questionOptions.concept && field.id.startsWith(question.id),
+    const repeatedFields = allFormFields.filter(
+      (field) =>
+        field.questionOptions.concept === question.questionOptions.concept &&
+        field.id.startsWith(question.id) &&
+        !field.meta?.repeat?.wasDeleted,
+    );
+    setCounter(repeatedFields.length - 1);
+    setRows(repeatedFields);
```

Comment:
Can these values be [computed](https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state) during rendering? Does this work as well as the current logic?

```tsx
  const { fields: allFormFields, encounterContext } = React.useContext(FormContext);
  const repeatedFields = allFormFields.filter(
    (field) =>
      field.questionOptions.concept === question.questionOptions.concept &&
      field.id.startsWith(question.id) &&
      !field.meta?.repeat?.wasDeleted,
  );

  const [counter, setCounter] = useState(repeatedFields.length - 1 ?? 0);
  const [rows, setRows] = useState(repeatedFields ?? []);

 // ... other code omitted for brevity

  useEffect(() => {
     getFieldControlWithFallback(getQuestionWithSupportedRendering(question))?.then((component) =>
      setFieldComponent({ Component: component }),
    );
  }, [allFormFields, question]);
```

Reply:
Not sure, @samuelmale. I haven't tested that snippet out.

Reply:
Not that I've done any kind of smell test but by look of things it wouldn't sync with the counter state, not hard to fix. Do you have reservations with the current state of things so that I consider refactoring the component or this was just a nice to have thing?

Reply:
Nice to have